source/credentials_provider_process.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/auth/credentials.h>
+#include <aws/auth/external/cJSON.h>
+#include <aws/auth/private/aws_profile.h>
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/common/clock.h>
+#include <aws/common/date_time.h>
+#include <aws/common/environment.h>
+#include <aws/common/process.h>
+#include <aws/common/string.h>
+
+#if defined(_MSC_VER)
+#    pragma warning(disable : 4204)
+#endif /* _MSC_VER */
+
+struct aws_credentials_provider_process_impl {
+    struct aws_string *command;
+};
+
+AWS_STATIC_STRING_FROM_LITERAL(s_access_key_id_name, "AccessKeyId");
+AWS_STATIC_STRING_FROM_LITERAL(s_secret_access_key_name, "SecretAccessKey");
+AWS_STATIC_STRING_FROM_LITERAL(s_session_token_name, "Token");
+AWS_STATIC_STRING_FROM_LITERAL(s_creds_expiration_name, "Expiration");
+
+static struct aws_credentials *s_parse_credentials_from_json_document(
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *document) {
+
+    struct aws_credentials *credentials = NULL;
+    cJSON *document_root = NULL;
+    bool success = false;
+    bool parse_error = true;
+
+    document_root = cJSON_Parse((const char *)document->buffer);
```

Comment:
Do the keys here match any of the other JSON doc providers? Any way to refactor such that we only do this once?  I feel like I've seen this function 2-3 times.  Maybe something that takes a JSON doc, the key of the primary section where everything is and then the 3 keys used to pull credential ccomponents?

Reply:
I can rework this function to be a shared one.